components/http/okHttp/src/main/java/com/microsoft/kiota/http/middleware/AuthorizationHandler.java

```diff
@@ -0,0 +1,127 @@
+
+    /**
+     * Instantiates a new AuthorizationHandler.
+     * @param authenticationProvider the authentication provider.
+     */
+    public AuthorizationHandler(BaseBearerTokenAuthenticationProvider authenticationProvider) {
+        this.authenticationProvider = authenticationProvider;
+    }
+
+    @Override
+    public Response intercept(final Chain chain) throws IOException {
+        Objects.requireNonNull(chain, "parameter chain cannot be null");
+        final Request request = chain.request();
+
+        final Span span =
+                ObservabilityHelper.getSpanForRequest(request, "AuthorizationHandler_Intercept");
+        Scope scope = null;
+        if (span != null) {
+            scope = span.makeCurrent();
+            span.setAttribute("com.microsoft.kiota.handler.authorization.enable", true);
+        }
+
+        try {
+            // Auth provider already added auth header
+            if (request.headers().names().contains(authorizationHeaderKey)) {
+                span.setAttribute("com.microsoft.kiota.handler.authorization.token_present", true);
+                return chain.proceed(request);
+            }
+
+            authenticateRequest(request, null, span);
+            Response response = chain.proceed(chain.request());
+
+            if (response != null && response.code() != HttpURLConnection.HTTP_UNAUTHORIZED) {
+                return response;
+            }
+
+            // Attempt CAE claims challenge
+            String claims = ContinuousAccessEvaluationClaims.getClaimsFromResponse(response);
+            if (claims == null || claims.isEmpty()) {
+                return response;
+            }
+
+            span.addEvent("com.microsoft.kiota.handler.authorization.challenge_received");
+
+            // We cannot replay one-shot requests after claims challenge
+            boolean isRequestBodyOneShot =
+                    request != null && request.body() != null && request.body().isOneShot();
+            if (isRequestBodyOneShot) {
+                return response;
+            }
+
+            response.close();
+            final HashMap<String, Object> additionalContext = new HashMap<>();
+            additionalContext.put("claims", claims);
+            // Retry claims challenge only once
+            authenticateRequest(request, additionalContext, span);
+            span.setAttribute(HTTP_REQUEST_RESEND_COUNT, 1);
+            return chain.proceed(request);
+        } finally {
+            if (scope != null) {
+                scope.close();
+            }
+            if (span != null) {
+                span.end();
+            }
+        }
+    }
+
+    private void authenticateRequest(
+            @Nonnull Request request,
+            @Nullable Map<String, Object> additionalAuthenticationContext,
+            @Nonnull Span span) {
+        final RequestInformation requestInformation = getRequestInformation(request);
+        authenticationProvider.authenticateRequest(
```

Comment:
this should get the access token provider from the base authentication provider instead so we don't have to create an intermediate request information object.